### PR TITLE
chore: remove middleware to request version and entitlement warnings

### PR DESCRIPTION
**Pull request by kylecarbs**
This cleans up `root.go` a bit, adds tests for middleware HTTP transport functions, and removes two HTTP requests we have always performed previously when executing *any* client command. It should improve CLI performance (especially for users with higher latency).

Version information was already sent as a header to every request. This adds entitlement warnings to requests if they exist. This seems reasonable because it's rare for users to have entitlement warnings.

The only behavior this breaks is when running `coder logout`; if you're already logged out, it will no longer error.

I removed a test that regressed in [this commit](https://github.com/coder/coder/commit/d2998c6b7b9fe9f3d7922108dab536dbf08d6fc5#diff-e1d0a50ee0a7500620f021bf79f28ffe6f0fc754c08ce01f8a07d8d5f7fd50a2). It's for the deprecated command, so it seems minimal risk since it wasn't working anyways.

### CODE REVIEW

File: cli/root.go [hunk 1]
```diff
@@ -67,8 +67,7 @@ const (
 	varOrganizationSelect = "organization"
 	varDisableDirect      = "disable-direct-connections"
 
-	notLoggedInMessage         = "You are not logged in. Try logging in using 'coder login <url>'."
-	notLoggedInURLSavedMessage = "You are not logged in. Try logging in using 'coder login'."
+	notLoggedInMessage = "You are not logged in. Try logging in using 'coder login <url>'."
```
Comment (mafredri) [inline cli/root.go:70]:
```suggestion
	notLoggedInMessage = "You are not logged in. Try logging in using 'coder login [url]'."
```
URL is sometimes optional, if we're only using on message then I'd prefer highlighting this.

File: cli/root.go [hunk 2]
```diff
@@ -700,22 +589,39 @@ func (r *RootCmd) HeaderTransport(ctx context.Context, serverURL *url.URL) (*cod
 	return transport, nil
 }
 
-func (r *RootCmd) setClient(ctx context.Context, client *codersdk.Client, serverURL *url.URL) error {
-	transport, err := r.HeaderTransport(ctx, serverURL)
+// configureClient
```
Comment (coadler) [inline cli/root.go:592]: rm comment or add something to it

File: cli/root.go [hunk 3]
```diff
@@ -1305,3 +1134,146 @@ func defaultUpgradeMessage(version string) string {
...
+		return res, err
+	})
+}
+
+// wrapTransportWithTelemetryHeader adds telemetry headers to report command usage
+// to an HTTP transport.
+func wrapTransportWithTelemetryHeader(transport http.RoundTripper, inv *serpent.Invocation) http.RoundTripper {
+	var (
+		mu    sync.RWMutex
+		done  bool
```
Comment (coadler) [inline cli/root.go:1224]: Couldn't this just be a `sync.Once`?

File: cli/root.go [hunk 4]
```diff
@@ -1305,3 +1134,146 @@ func defaultUpgradeMessage(version string) string {
 	}
 	return fmt.Sprintf("download the server version with: 'curl -L https://coder.com/install.sh | sh -s -- --version %s'", version)
 }
+
+// wrapTransportWithEntitlementsCheck adds a middleware to the HTTP transport
+// that checks for entitlement warnings and prints them to the user.
+func wrapTransportWithEntitlementsCheck(rt http.RoundTripper, w io.Writer) http.RoundTripper {
+	var (
+		mu   sync.RWMutex
```
Comment (mafredri) [inline cli/root.go:1142]: I think these functions could be simplified with a plain mutex (vs rw). These code paths aren't really hot enough to benefit from read locking IMO.

**Review by coadler — APPROVED** [review]

File: cli/templatecreate_test.go [hunk 5]
```diff
@@ -197,56 +196,6 @@ func TestTemplateCreate(t *testing.T) {
 		require.NoError(t, err, "Template must be recreated without error")
 	})
 
-	t.Run("WithVariablesFileWithoutRequiredValue", func(t *testing.T) {
```
Comment (coadler) [inline cli/templatecreate_test.go:200]: Also, this seems unrelated. Why rm?

File: cli/root.go [hunk 6]
```diff
@@ -1305,3 +1134,146 @@ func defaultUpgradeMessage(version string) string {
...
+		serverInfo, err := getBuildInfo(inv.Context())
+		if err == nil && serverInfo.UpgradeMessage != "" {
+			upgradeMessage = serverInfo.UpgradeMessage
+		}
+		fmtWarningText := "version mismatch: client %s, server %s\n%s"
+		fmtWarn := pretty.Sprint(cliui.DefaultStyles.Warn, fmtWarningText)
+		warning := fmt.Sprintf(fmtWarn, clientVersion, serverVersion, upgradeMessage)
+
+		_, _ = fmt.Fprint(inv.Stderr, warning)
+		_, _ = fmt.Fprintln(inv.Stderr)
```
Comment (mafredri) [inline cli/root.go:1212]:
```suggestion
		_, _ = fmt.Fprintln(inv.Stderr, warning)
```

File: cli/templatecreate_test.go (hunk 5 above)
Comment (kylecarbs) [reply cli/templatecreate_test.go:200]: check pr description

**Review by mafredri — APPROVED** [review]

File: cli/root.go (hunk 1 above)
Comment (kylecarbs) [reply cli/root.go:70]: Well the funny thing is for you to get this message, you have to already hit the server.

File: cli/root.go (hunk 4 above)
Comment (kylecarbs) [reply cli/root.go:1142]: Fair I suppose this is just for client requests.

File: cli/templatecreate_test.go (hunk 5 above)
Comment (coadler) [reply cli/templatecreate_test.go:200]: oopsie

File: cli/root.go (hunk 3 above)
Comment (kylecarbs) [reply cli/root.go:1224]: Oop, yes

File: cli/root.go (hunk 1 above)
Comment (kylecarbs) [reply cli/root.go:70]: It seems like this shows the URL if you don't have it, and `coder login` if you do have it. That code wasn't really used properly!
